app/models/ability.rb

```diff
@@ -11,12 +13,15 @@ class Ability
   include CanCan::Ability
 
   def initialize(user)
-    @user = user
+    @user = user.present? ? user.decorate.object : nil
```

Comment:
This is weird. Why do we need to decorate this inside the ability definition?

Reply:
Because when we do `authorize! :whetever, Object`, or `can? :whetever, Object` it internally does for us something like this:
`Ability.new(current_user).can?(:whetever, Object)`
And the problem with that is that `current_user` is decorated and our permissions are for User model, not a UserDecorator object... so it will say that this user does not have permission although it does.

Reply:
If you want to see an example of what I am saying, do this:

1. Create `test/integration/provider/admin/user/personal_details_controller_test.rb` with this code:
```
# frozen_string_literal: true

require 'test_helper'

class Provider::Admin::User::PersonalDetailsControllerTest < ActionDispatch::IntegrationTest
  def setup
    provider = FactoryBot.create(:provider_account)
    @user = FactoryBot.create(:active_user, account: provider)
    login! provider, user: user
  end

  attr_reader :user

  NO_NOTIFICATIONS_REGEX = /You don't have access to any notifications/

  test 'No notification preferences available' do
    user.admin_sections = []
    user.save!(validate: false)

    get provider_admin_user_notification_preferences_path

    page = Nokogiri::HTML::Document.parse(response.body)
    assert_match NO_NOTIFICATIONS_REGEX, page.text
  end

  test 'Some notification preferences available' do
    user.admin_sections = ['partners']
    user.save!(validate: false)

    get provider_admin_user_notification_preferences_path

    page = Nokogiri::HTML::Document.parse(response.body)
    assert_not_match NO_NOTIFICATIONS_REGEX, page.text
  end
end
```

2. Run it, it will fail without this change.
In this case it is for the authorize in `app/controllers/provider/admin/user/base_controller.rb`.
But keep in mind that this is ONE individual scenario, but we also have plenty of `can? ...` inside views.

Reply:
I see. It's actually to "undecorate" it. It doesn't look right anyway. `Ability` expects an `User` but what we pass may not be an `User` yet.

Reply:
> Why do we need to decorate this inside the ability definition?

Btw, we are not decorating here, but _undecorating_ instead. If it is a user model or its decorated object, it still saves the model in `@user` and not the decorated version.

Reply:
> I see. It's actually to "undecorate" it. It doesn't look right anyway. Ability expects a User but what we pass may not be a User yet.

But what do you propose? 😕

Reply:
The alternative version that I could think of is checking where we do `Ability.new` and THERE sending the current_user without being decorated... WDYT?